### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry): smooth morphisms of schemes

### PR DESCRIPTION
**Pull request by chrisflav**
We say a morphism of schemes is smooth (of relative dimension `n`) if for every point of the source there exist affine open neighborhoods of the point and its image such that the induced ring map is standard smooth (of relative dimension `n`). These notions are local on the source and the target and satisfy the standard stability properties.

This contribution was created as part of the AIM workshop "Formalizing algebraic geometry" in June 2024.

---

- [x] depends on: #16868
- [x] depends on: #17305 
- [x] depends on: #17335 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Presentation.lean [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+generators with
+1. `relations`: The type of relations.
+2. `relation : relations → MvPolynomial vars R`: The assignment of
+each relation to a polynomial in the generators.
+-/
+structure Algebra.Presentation extends Algebra.Generators.{w} R S where
+  /-- The type of relations.  -/
+  relations : Type t
+  /-- The assignment of each relation to a polynomial in the generators. -/
+  relation : relations → MvPolynomial vars R
```
[the remark this answers is not in the document]
Comment (chrisflav) [reply Mathlib/RingTheory/Presentation.lean:44]: That syntax does not seem to work? It gives "unkown identifier `P`"

File: Mathlib/RingTheory/Presentation.lean [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  toFun := P.fromQuotient
+  invFun s := P.σ s
+  left_inv x := Quotient.inductionOn x <| fun x ↦
+    (Ideal.Quotient.mk_eq_mk_iff_sub_mem _ _).mpr
+    (P.σ'_aeval_val_sub_mem_span_range_relation x)
+  right_inv x := by simp
+  map_mul' := by simp
+  map_add' := by simp
+  commutes' := by simp
+
```
[the remark this answers is not in the document]
Comment (chrisflav) [reply Mathlib/RingTheory/Presentation.lean]: I think we should have the equivalence, but I refactored it so its better now.

File: Mathlib/RingTheory/Presentation.lean [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+-/
+structure Algebra.Presentation extends Algebra.Generators.{w} R S where
+  /-- The type of relations.  -/
+  relations : Type t
+  /-- The assignment of each relation to a polynomial in the generators. -/
+  relation : relations → MvPolynomial vars R
+  /-- The relations are mapped to zero under the canonical map. -/
+  relation_zero (r : relations) : MvPolynomial.aeval val (relation r) = 0
+  σ'_aeval_val_sub_mem_span_range_relation (p : MvPolynomial vars R) :
+    σ' ((aeval val) p) - p ∈ Ideal.span (Set.range <| relation)
```
[the remark this answers is not in the document]
Comment (chrisflav) [reply Mathlib/RingTheory/Presentation.lean]: I agree, that's cleaner. Done.

**Comment by chrisflav** [comment]
> @chrisflav . It seems that you are planning to split this up into multiple PRs. What is the status of that? It would be great to get this work moving along!

Sorry, I was pushing on other fronts lately and did not have time to come to the review party. The first two PRs in this series are merged:
- #14271 
- #14325 
And I'll continue this series after my summer vacation.

**Comment by grunweg** [comment]
Enjoy your vacation! I'm labelling this as awaiting-author then.

File: Mathlib/RingTheory/RingHomProperties.lean [hunk 4]
```diff
@@ -167,6 +165,45 @@ theorem StableUnderBaseChange.pushout_inl (hP : RingHom.StableUnderBaseChange @P
...
+    (f : R →+* S) [IsLocalization (M.map f) Sᵣ] (hf : P f) :
+    P (IsLocalization.map Sᵣ f M.le_comap_map : Rᵣ →+* Sᵣ) := by
+  letI : Algebra Rᵣ Sᵣ := (IsLocalization.map Sᵣ f M.le_comap_map).toAlgebra
+  letI : Algebra R S := f.toAlgebra
+  letI : Algebra R Sᵣ := ((algebraMap Rᵣ Sᵣ).comp (algebraMap R Rᵣ)).toAlgebra
+  haveI : IsScalarTower R S Sᵣ := IsScalarTower.of_algebraMap_eq'
+    (IsLocalization.map_comp M.le_comap_map)
+  haveI : IsScalarTower R Rᵣ Sᵣ := IsScalarTower.of_algebraMap_eq' rfl
+  haveI : IsLocalization (Algebra.algebraMapSubmonoid S M) Sᵣ :=
+    inferInstanceAs <| IsLocalization (M.map f) Sᵣ
```
Comment (jcommelin) [inline Mathlib/RingTheory/RingHomProperties.lean:204]: Can this be golfed with `algebraize`?

**Review by jcommelin — COMMENTED** [review]
Nice milestone!

File: Mathlib/RingTheory/RingHomProperties.lean (hunk 4 above)
Comment (chrisflav) [reply Mathlib/RingTheory/RingHomProperties.lean]: Yes, and I had already done this in some other PR :D I had messed up the merging. Is now fixed.

**Review by jcommelin — COMMENTED** [review]
maintainer merge

File: Mathlib/AlgebraicGeometry/Morphisms/Smooth.lean [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+open CategoryTheory
+
+universe t w v u
+
+namespace AlgebraicGeometry
+
+open RingHom
+
+variable (n m : ℕ)
+variable {X Y : Scheme.{u}} (f : X ⟶ Y)
```
Comment (riccardobrasca) [inline Mathlib/AlgebraicGeometry/Morphisms/Smooth.lean:53]: I guess you can use only one line here. (Not very important of course.)

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

**Review by erdOne — APPROVED** [review]

File: Mathlib/AlgebraicGeometry/Morphisms/Smooth.lean [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+/--
+A morphism of schemes `f : X ⟶ Y` is smooth if for each `x : X` there
+exists an affine open neighborhood `V` of `x` and an affine open neighborhood `U` of
+`f.base x` with `V ≤ f ⁻¹ᵁ U` such that the induced map `Γ(Y, U) ⟶ Γ(X, V)` is
+standard smooth.
+-/
+@[mk_iff]
+class IsSmooth : Prop where
+  exists_isStandardSmooth : ∀ (x : X), ∃ (U : Y.affineOpens) (V : X.affineOpens) (_ : x ∈ V.1)
+    (e : V.1 ≤ f ⁻¹ᵁ U.1), IsStandardSmooth.{0, 0} (f.appLE U V e)
```
[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Morphisms/Smooth.lean:63]: Or maybe we should just drop the universe argument from `IsStandardSmooth` completely and define a constructor that takes arbitrary universe presentation instead.

[the remark this answers is not in the document]
Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Morphisms/Smooth.lean:63]: Gives less nice def-eqs though if you are working with explicit presentations of rings.

[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Morphisms/Smooth.lean:63]: Wait does `IsStandardSmooth` contain data?

[the remark this answers is not in the document]
Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Morphisms/Smooth.lean:63]: Ah sorry, I was confused and thought you were talking about `Presentation`. I agree, dropping universe parameters from `IsStandardSmooth` is fine.

**Comment by chrisflav** [comment]
Thanks for all the reviews!

bors r+

**Pull request closed without merging**